src/os/bluestore/ZonedFreelistManager.cc

```diff
@@ -0,0 +1,325 @@
+  if (enumerate_zone_num == ~0UL) {
+    enumerate_p = kvdb->get_iterator(info_prefix);
+    enumerate_p->lower_bound(string());
+    enumerate_zone_num = 0;
+  } else {
+    enumerate_p->next();
+    if (!enumerate_p->valid()) {
+      dout(10) << __func__ << " end" << dendl;
+      return false;
+    }
+    ++enumerate_zone_num;
+  }
+
+  zone_state_t zone_state(enumerate_p, zone_size);
+  ceph_assert(zone_state.get_zone_num() == enumerate_zone_num);
+
+  *offset = zone_state.get_offset();
+  *length = zone_state.get_remaining_space();
+
+  return true;
+}
+
+void ZonedFreelistManager::dump(KeyValueDB *kvdb) {
+  enumerate_reset();
+  uint64_t offset, length;
+  while (enumerate_next(kvdb, &offset, &length)) {
+    dout(20) << __func__ << " 0x" << std::hex << offset << "~" << length
+	     << std::dec << dendl;
+  }
+}
+
+// Advances the write pointer and writes the updated write pointer to database.
+void ZonedFreelistManager::allocate(
+    uint64_t offset,
+    uint64_t length,
+    KeyValueDB::Transaction txn) {
+  dout(10) << __func__ << " 0x" << std::hex << offset << "~" << length << dendl;
+  uint64_t zone_num = offset / zone_size;
+  zone_state_t zone_state(zone_num, zone_size);
+  zone_state.increment_write_pointer(length);
+  zone_state.persist(info_prefix, txn);
+}
+
+// Increments the number of dead bytes in a zone and writes the updated value to
+// database.  The dead bytes in the zone are not usable.  The cleaner will later
+// copy live objects from the zone to another zone an make the zone writable
+// again.  The number of dead bytes in a zone is used by the cleaner to select
+// which zones to clean -- the ones with most dead bytes are good candidates
+// since they require less I/O.
+void ZonedFreelistManager::release(
+    uint64_t offset,
+    uint64_t length,
+    KeyValueDB::Transaction txn) {
+  dout(10) << __func__ << " 0x" << std::hex << offset << "~" << length << dendl;
+  uint64_t zone_num = offset / zone_size;
+  zone_state_t zone_state(zone_num, zone_size);
+  zone_state.increment_num_dead_bytes(length);
+  zone_state.persist(info_prefix, txn);
+}
+
+void ZonedFreelistManager::get_meta(
+    uint64_t target_size,
+    std::vector<std::pair<string, string>>* res) const {
+  // We do not support expanding devices for now.
+  ceph_assert(target_size == 0);
+  res->emplace_back("zfm_size", stringify(size));
+  res->emplace_back("zfm_bytes_per_block", stringify(bytes_per_block));
+  res->emplace_back("zfm_zone_size", stringify(zone_size));
+  res->emplace_back("zfm_num_zones", stringify(num_zones));
+  res->emplace_back("zfm_starting_zone_num", stringify(starting_zone_num));
+}
+
+std::vector<zone_state_t> ZonedFreelistManager::get_zone_states(
+    KeyValueDB *kvdb) const {
+  std::lock_guard l(lock);
+  std::vector<zone_state_t> zone_states;
+  auto p = kvdb->get_iterator(info_prefix);
+  for (p->lower_bound(string()); p->valid(); p->next()) {
+    zone_states.push_back(zone_state_t(p, zone_size));
```

Comment:
Are you suggesting get_zone_states to take as an argument the range of zones to load?  Regarding dependence, I don't see how zone_states_t is allocator-dependent.

Perhaps you can sketch out what you have in mind, and I will implement it.

Reply:
I meant something like passing a callback (targeting to allocator) which to be called when a portion of zones is loaded.
But honestly I expected much smaller zones (something like ten MBs) and hence much longer zone list....

Reply:
If you don't mind, I will leave it as is for now because there is only one client of get_zone_states(), which runs at startup -- i.e. it is not on a critical path.  We can revisit later if need be.

Reply:
sounds good.